### PR TITLE
Better typing of model attributes.

### PR DESCRIPTION
**Pull request by geon**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://www.typescriptlang.org/docs/handbook/advanced-types.html (This makes a lot of the setters and getters typed.)
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by geon** [comment]
I haven't updated the test, so it fails, obviously.

The Backbone.Model class needs to be instantiated with a generic parameter, so this:

´´´js
class User extends Backbone.Model {
    email: string,
    name: string
}
´´´

...becomes:

´´´js
class User extends Backbone.Model<{
    email: string,
    name: string
}>
´´´
You will have to decide if you think it is worth the hazzle.

File: types/backbone/index.d.ts [hunk 1]
```diff
@@ -101,20 +107,19 @@ declare namespace Backbone {
 
     class ModelBase extends Events {
         url: any;
-        parse(response: any, options?: any): any;
         toJSON(options?: any): any;
         sync(...arg: any[]): JQueryXHR;
     }
 
-    class Model extends ModelBase {
+    class Model<TAttributes extends {}> extends ModelBase {
```
Comment (minestarks) [inline types/backbone/index.d.ts]: I don't think you want `extends {}` here, as everything extends `{}` (even `string` or `number`) so you may as well omit it. Perhaps `object` is what you want if you want to exclude primitive types.

File: types/backbone/index.d.ts [hunk 2]
```diff
@@ -129,35 +134,23 @@ declare namespace Backbone {
...
-        /**
-        * For strongly-typed assignment of attributes, use the `set` method only privately in public setter properties.
-        * @example
-        * set name(value: string) {
-        *    super.set("name", value);
-        * }
-        **/
-        /*private*/ set(attributeName: string, value: any, options?: ModelSetOptions): Model;
-        set(obj: any, options?: ModelSetOptions): Model;
+        // TODO: this still doesn't type check!!!
```
Comment (minestarks) [inline types/backbone/index.d.ts]: Could you update this comment to be more explanatory?

**Review by minestarks — CHANGES_REQUESTED** [review]

**Comment by minestarks** [comment]
authors @borisyankov @nvivo any comments?

@geon the tests will need to be fixed before we can merge.

File: types/backbone/index.d.ts (hunk 2 above)
Comment (geon) [reply types/backbone/index.d.ts]: Sorry, I don't know why the typing on that method is broken. It *should* work...

File: types/backbone/index.d.ts (hunk 1 above)
Comment (geon) [reply types/backbone/index.d.ts]: Right.

File: types/backbone/index.d.ts (hunk 2 above)
Comment (minestarks) [reply types/backbone/index.d.ts]: I might be able to help - can you give an example of how it's broken?

Comment (geon) [reply types/backbone/index.d.ts]: When I tried it, `set` behaved like it was completely untyped. Perhaps it was a bug in the version of Typescript at the time. I haven't used this code for a couple of months.

**Comment by minestarks** [comment]
@geon Just confirming you still pursuing this PR? If you don't intend to make follow up changes I can close it out and you can open a new one once you have more time.

**Comment by geon** [comment]
Not really interested anymore since I switched to React.

I just dumped this here in case someone else is interested.

**Comment by andy-ms** [comment]
Closing this since it won't be updated. @nvivo and @borisyankov may be interested in taking inspiration from it.

**Pull request closed without merging**